### DISCUSSION TITLE
Why is LFS failing when reverse lookup is not working

### DISCUSSION THREAD

**Discussion by MarRich1** [git-lfs/git-lfs#5833] [Q&A]
Hi,

we are using Azure DevOps Server in our intranet. 
Last time we were facing issues, when the network team disabled reverse lookup. The clients wasn't able anymore to authenticate to the LFS share, which is the AzDO server itself.

Why is the reverse lookup needed, and is there a way to change this?

Our security department is the opinion, that a reverse lookup is an security issue

**Comment by chrisd8088** [comment]
Hey, I'm sorry you're having trouble.  We would need a lot more information to try to give you any advice, I think, and I would also suggest you contact the Azure DevOps support team, as this project is just for the open-source Git LFS client.

I assume you're asking about reverse lookup of a DNS record from an IP address.  Can you provide some further details?  Do you have a trace log (generated with `GIT_TRACE=1 GIT_TRANSFER_TRACE=1 GIT_CURL_VERBOSE=1`) or some other specifics regarding the points at which the Git LFS performs this lookup?

**Reply by MarRich1** [reply]
Does this help:

10:12:01.037008 trace git-lfs: HTTP: POST
https://server1.domain.local/tfs/Projects/projectname/_git/MyRepo.git/info/lfs/objects/batch
10:12:01.045267 trace git-lfs: http: got Negotiate error: lookup server1.domain.local: getaddrinfow: The requested name is valid, but no data of the requested type was found.
10:12:01.045267 trace git-lfs: setting repository access to basic
10:12:01.047781 trace git-lfs: exec: git 'config' '--includes' '--replace-all' 'lfs.
https://server1.domain.local/tfs/Projects/projectname/_git/MyRepo.git/info/lfs.access' 'basic'
10:12:01.869445 trace git-lfs: api: http response indicates "basic" authentication. Resubmitting...
10:12:01.869445 trace git-lfs: creds: git credential fill ("https", "server1.domain.local", "")
10:12:01.871590 trace git-lfs: exec: git 'credential' 'fill'
10:12:03.464120 trace git-lfs: Filled credentials for
https://server1.domain.local/tfs/Projects/projectname/_git/MyRepo
10:12:03.465031 trace git-lfs: HTTP: POST
https://server1.domain.local/tfs/Projects/projectname/_git/MyRepo.git/info/lfs/objects/batch
10:12:04.166708 trace git-lfs: HTTP: 401
10:12:04.171204 trace git-lfs: exec: git 'credential' 'reject'
10:12:04.906345 trace git-lfs: api: http response indicates "basic" authentication. Resubmitting...
10:12:04.906886 trace git-lfs: creds: git credential fill ("https", "server1.domain.local", "")
10:12:04.909061 trace git-lfs: exec: git 'credential' 'fill'
10:12:05.590831 trace git-lfs: Filled credentials for
https://server1.domain.local/tfs/Projects/projectname/_git/MyRepo
10:12:05.590831 trace git-lfs: HTTP: POST
https://server1.domain.local/tfs/Projects/projectname/_git/MyRepo.git/info/lfs/objects/batch
10:12:06.291342 trace git-lfs: HTTP: 401

**Reply by chrisd8088** [reply]
Thanks for the trace log!  I see the error message from the `getaddrinfow` API call, but that would be a normal DNS lookup of an IP address, I believe, not a reverse one of a DNS record from an IP address.

What is the address shown in the `Endpoint` line of the output of `git lfs env`, when you run that command in the working tree of your repository?

I was sort of expecting, from your description, that you were doing something like `git clone https://192.168.1.42/Projects/projectname`, and then seeing the Git LFS client trying to figure out what the domain name associated with that IP address was, and that was failing.  But it seems like the client has a domain name, and is just getting a 401 when making a request to it, which looks more like the kind of error we often see with firewalls and proxies that interfere with the usual HTTP authentication process.

I think we'd need more detail about your situation, though, to say anything else.

**Comment by MarRich1** [comment]
The “Git server” we run is an Azure DevOps server that enforces Kerberos authentication instead of NTLM.
And yes, of course we have a “well-functioning” network infrastructure in place.
As long as reverse lookup works, LFS works fine, but our security department would like to disable reverse lookup, and I would like to understand how this affects GIT LFS.

Unfortunately, there is no official company that I can throw money at to analyze the problem

With best regards,
Markus

**Reply by chrisd8088** [reply]
OK -- well, I don't know what else to advise other than what I said above.  The one error message in your trace log looks like it has something to do with forward DNS lookup, not reverse lookup:
```
lookup server1.domain.local: getaddrinfow: The requested name is valid, but no data of the requested type was found
```

What do the HTTP requests look like to the Git LFS Batch API?  Your client is making a request like the following; what are the request and response headers?  You can use `GIT_CURL_VERBOSE=1` to capture those.
```
https://server1.domain.local/tfs/Projects/projectname/_git/MyRepo.git/info/lfs/objects/batch
```

I don't suppose you have a proxy for the Kerberos authentication?  I'm reminded of #5401; perhaps that has some bearing here?

**Comment by MarRich1** [comment]
Sorry for the late response. Yes this seems to be the same issue, but the problem also occurs when w/o a proxy
The problematic part seems to be the imported module https://github.com/dpotapov/go-spnego/blob/master/spnego.go which makes an additional check if hostname <=> ip address matches. 

Maybe an additional switch in the git config can avoid this behavior. 

eg: https://github.com/git-lfs/git-lfs/blob/main/lfshttp/client.go line 496 `return &spnego.Transport{Transport: *tr}, nil` with the `NoCanonicalize`parameter depending on the config file

**Reply by chrisd8088** [reply]
No worries at all about taking time to investigate something; thank you very much!  I think that does sound like a probably culprit here for the behaviour you've described.  Specifically, as you say, the `dpotapov/go-spnego` package's `canonicalizeHostname()` method [invokes](https://github.com/dpotapov/go-spnego/blob/298b63a544303a239753d04314aada5bdbad7e4a/spnego.go#L23) `net.LookupAddr()` in its `canonicalizeHostname()` function, which it [calls](https://github.com/dpotapov/go-spnego/blob/298b63a544303a239753d04314aada5bdbad7e4a/spnego_windows.go#L20) in either the Windows or non-Windows [versions](https://github.com/dpotapov/go-spnego/blob/298b63a544303a239753d04314aada5bdbad7e4a/spnego_gokrb5.go#L69) of the `SetSPNEGOHeader()` method it [calls](https://github.com/dpotapov/go-spnego/blob/298b63a544303a239753d04314aada5bdbad7e4a/transport.go#L27) in its implementation of the `RoundTrip()` method of the `http.Transport` interface.  And we [establish](https://github.com/git-lfs/git-lfs/blob/fc61febe9cc2d9ddc6ffe3e8d1ae546512632552/lfshttp/client.go#L496) a `spnego.Transport` when we're dealing with Kerberos authentication.

The links above are for the version of `dpotapov/go-spnego` we're [using](https://github.com/git-lfs/git-lfs/blob/fc61febe9cc2d9ddc6ffe3e8d1ae546512632552/go.sum#L9-L10) now, so to take advantage of the optional hostname canonicalization introduced in PR dpotapov/go-spnego#5, we would need to update to a newer version of the package

We could then introduce a Git LFS configuration which would disable the canonicalization, as you suggest, or conceivably just disable it for everyone.  The configuration setting approach is probably safer; maybe we could add a setting named `lfs.negotiateAccess.canonicalizeHostname` which is `true` by default, and which we then negate and pass as the `NoCanonicalize` parameter to a new `spnego.Transport` structure.

If anyone is able to write up a patch (ideally with a test, although I'm not sure offhand to make that work) and propose it in a PR, I'd be delighted to take a look.  Otherwise, I'll mark this as a possible enhancement and we'll try to keep it in our backlog.

Thanks again for the diagnosis!

**Reply by chrisd8088** [reply]
I've opened #5846 to track this as an issue.  Thanks again for the report and the investigation.
